crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn is_regular_package(&self, db: &dyn Db, search_path: impl Into<Self>) -> bool {
+        match (self, search_path.into()) {
+            (Self::Extra(ExtraPath(fs_path)), Self::Extra(_))
+            | (Self::FirstParty(FirstPartyPath(fs_path)), Self::FirstParty(_))
+            | (Self::SitePackages(SitePackagesPath(fs_path)), Self::SitePackages(_)) => {
+                let file_system = db.file_system();
+                file_system.exists(&fs_path.join("__init__.py"))
+                    || file_system.exists(&fs_path.join("__init__.pyi"))
+            }
+            // Unlike the other variants:
+            // (1) Account for VERSIONS
+            // (2) Only test for `__init__.pyi`, not `__init__.py`
+            (Self::StandardLibrary(StandardLibraryPath(fs_path)), Self::StandardLibrary(stdlib_root)) => {
+                let Some(module_name) = self.as_module_name() else {
+                    return false;
+                };
+                let typeshed_versions = Self::load_typeshed_versions(db, stdlib_root);
+                match typeshed_versions.query_module(&module_name, get_target_py_version(db)) {
+                    TypeshedVersionsQueryResult::Exists
+                    | TypeshedVersionsQueryResult::MaybeExists => {
+                        db.file_system().exists(&fs_path.join("__init__.pyi"))
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => false,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn parent(&self) -> Option<Self> {
+        Some(match self {
+            Self::Extra(ExtraPath(path)) => Self::extra_unchecked(path.parent()?),
+            Self::FirstParty(FirstPartyPath(path)) => Self::first_party_unchecked(path.parent()?),
+            Self::StandardLibrary(StandardLibraryPath(path)) => {
+                Self::standard_library_unchecked(path.parent()?)
+            }
+            Self::SitePackages(SitePackagesPath(path)) => {
+                Self::site_packages_unchecked(path.parent()?)
+            }
+        })
+    }
+
+    #[must_use]
+    fn ends_with_dunder_init(&self) -> bool {
+        match self {
+            Self::Extra(ExtraPath(path))
+            | Self::FirstParty(FirstPartyPath(path))
+            | Self::SitePackages(SitePackagesPath(path)) => {
+                path.ends_with("__init__.py") || path.ends_with("__init__.pyi")
+            }
+            Self::StandardLibrary(StandardLibraryPath(path)) => path.ends_with("__init__.pyi"),
+        }
+    }
+
+    #[must_use]
+    fn sans_dunder_init(self) -> Self {
```

Comment:
Nit: I would be calling this `strip_dunder_init` to align with `str::strip_sufix` (strip is probably also a more familiar word to none native English or French speakers)

Reply:
Pardon, monsieur